File: assets/js/liveview_react/uploads.ts
import type {
  LiveUploadConfig,
  LiveUploadEntry,
  LiveViewTarget,
  UploadFiles,
} from "./types";

const CONFIG_KEYS = Object.freeze([
  "accept",
  "auto_upload",
  "entries",
  "errors",
  "max_entries",
  "max_entries_mode",
  "max_file_size",
  "name",
  "ref",
]);

const ENTRY_KEYS = Object.freeze([
  "cancelled",
  "client_last_modified",
  "client_name",
  "client_relative_path",
  "client_size",
  "client_type",
  "done",
  "errors",
  "preflighted",
  "progress",
  "ref",
  "valid",
]);

const CONFIG_ERROR_KEYS = Object.freeze(["error", "ref"]);
const MAX_ENCODED_NESTING_DEPTH = 64;
const FORBIDDEN_ENCODED_KEYS = new Set([
  "__proto__",
  "constructor",
  "prototype",
]);

export interface NormalizedLiveUploadEntry extends LiveUploadEntry {
  readonly cancelled: boolean;
  readonly client_last_modified: number;
  readonly client_relative_path: string;
}

export interface NormalizedLiveUploadError {
  readonly error: unknown;
  readonly ref: string;
}

export interface NormalizedLiveUploadConfig extends LiveUploadConfig {
  readonly entries: readonly NormalizedLiveUploadEntry[];
  readonly errors: readonly NormalizedLiveUploadError[];
  readonly max_entries_mode: "selected" | "total";
  readonly max_file_size: number;
}

export interface ResolveLiveUploadInputOptions {
  readonly bridgeElement: HTMLElement | null;
  readonly changeEvent?: string;
  readonly formId?: string;
  readonly submitEvent?: string;
  readonly target?: LiveViewTarget;
}

export interface ResolvedLiveUploadInput {
  readonly form: HTMLFormElement;
  readonly input: HTMLInputElement;
}

function isRecord(value: unknown): value is Record<string, unknown> {
  if (typeof value !== "object" || value === null || Array.isArray(value)) {
    return false;
  }

  const prototype = Object.getPrototypeOf(value);
  return prototype === Object.prototype || prototype === null;
}

function assertExactKeys(
  value: Record<string, unknown>,
  expectedKeys: readonly string[],
  source: string,
): void {
  const expected = new Set(expectedKeys);
  const actualKeys = Object.keys(value);
  const unexpected = actualKeys.filter((key) => !expected.has(key));
  const missing = expectedKeys.filter(
    (key) => !Object.prototype.hasOwnProperty.call(value, key),
  );

  if (missing.length > 0 || unexpected.length > 0) {
    const details = [
      ...(missing.length > 0 ? [`missing ${missing.join(", ")}`] : []),
      ...(unexpected.length > 0 ? [`unexpected ${unexpected.join(", ")}`] : []),
    ].join("; ");
    throw new TypeError(`${source} has invalid fields: ${details}`);
  }
}

function assertNonEmptyString(value: unknown, source: string): string {
  if (typeof value !== "string" || value.length === 0) {
    throw new TypeError(`${source} must be a non-empty string`);
  }
  return value;
}

function assertString(value: unknown, source: string): string {
  if (typeof value !== "string") {
    throw new TypeError(`${source} must be a string`);
  }
  return value;
}

function assertBoolean(value: unknown, source: string): boolean {
  if (typeof value !== "boolean") {
    throw new TypeError(`${source} must be a boolean`);
  }
  return value;
}

function assertSafeInteger(
  value: unknown,
  source: string,
  minimum: number,
): number {
  if (!Number.isSafeInteger(value) || (value as number) < minimum) {
    const qualifier = minimum === 0 ? "non-negative" : "positive";
    throw new TypeError(`${source} must be a ${qualifier} safe integer`);
  }
  return value as number;
}

function cloneEncodedValue(
  value: unknown,
  source: string,
  ancestors: ReadonlySet<object> = new Set(),
  depth = 0,
): unknown {
  if (
    value === null ||
    typeof value === "string" ||
    typeof value === "boolean"
  ) {
    return value;
  }
  if (typeof value === "number") {
    if (!Number.isFinite(value)) {
      throw new TypeError(`${source} must contain only encoded JSON values`);
    }
    return value;
  }
  if (Array.isArray(value)) {
    if (depth >= MAX_ENCODED_NESTING_DEPTH) {
      throw new TypeError(
        `${source} exceeds the maximum encoded nesting depth of ${MAX_ENCODED_NESTING_DEPTH}`,
      );
    }
    if (ancestors.has(value)) {
      throw new TypeError(`${source} must not contain cyclic values`);
    }
    const nextAncestors = new Set(ancestors).add(value);
    return Object.freeze(
      value.map((item, index) =>
        cloneEncodedValue(
          item,
          `${source}[${index}]`,
          nextAncestors,
          depth + 1,
        ),
      ),
    );
  }
  if (isRecord(value)) {
    if (depth >= MAX_ENCODED_NESTING_DEPTH) {
      throw new TypeError(
        `${source} exceeds the maximum encoded nesting depth of ${MAX_ENCODED_NESTING_DEPTH}`,
      );
    }
    if (ancestors.has(value)) {
      throw new TypeError(`${source} must not contain cyclic values`);
    }
    const nextAncestors = new Set(ancestors).add(value);
    return Object.freeze(
      Object.fromEntries(
        Object.entries(value).map(([key, item]) => {
          if (FORBIDDEN_ENCODED_KEYS.has(key)) {
            throw new TypeError(
              `${source} contains forbidden object key "${key}"`,
            );
          }
          return [
            key,
            cloneEncodedValue(
              item,
              `${source}.${key}`,
              nextAncestors,
              depth + 1,
            ),
          ];
        }),
      ),
    );
  }

  throw new TypeError(`${source} must contain only encoded JSON values`);
}

function normalizeErrors(value: unknown, source: string): readonly unknown[] {
  if (!Array.isArray(value)) {
    throw new TypeError(`${source} must be an array`);
  }
  return Object.freeze(
    value.map((error, index) =>
      cloneEncodedValue(error, `${source}[${index}]`),
    ),
  );
}

function normalizeEntry(
  value: unknown,
  source: string,
): NormalizedLiveUploadEntry {
  if (!isRecord(value)) {
    throw new TypeError(`${source} must be an object`);
  }
  assertExactKeys(value, ENTRY_KEYS, source);

  const progress = assertSafeInteger(value.progress, `${source}.progress`, 0);
  if (progress > 100) {
    throw new TypeError(`${source}.progress must be between 0 and 100`);
  }

  return Object.freeze({
    cancelled: assertBoolean(value.cancelled, `${source}.cancelled`),
    client_last_modified: assertSafeInteger(
      value.client_last_modified,
      `${source}.client_last_modified`,
      0,
    ),
    client_name: assertString(value.client_name, `${source}.client_name`),
    client_relative_path: assertString(
      value.client_relative_path,
      `${source}.client_relative_path`,
    ),
    client_size: assertSafeInteger(
      value.client_size,
      `${source}.client_size`,
      0,
    ),
    client_type: assertString(value.client_type, `${source}.client_type`),
    done: assertBoolean(value.done, `${source}.done`),
    errors: normalizeErrors(value.errors, `${source}.errors`),
    preflighted: assertBoolean(value.preflighted, `${source}.preflighted`),
    progress,
    ref: assertNonEmptyString(value.ref, `${source}.ref`),
    valid: assertBoolean(value.valid, `${source}.valid`),
  });
}

function normalizeAccept(value: unknown): "any" | readonly string[] {
  if (value === "any") return "any";
  if (
    !Array.isArray(value) ||
    value.length === 0 ||
    !value.every((entry) => typeof entry === "string" && entry.length > 0)
  ) {
    throw new TypeError(
      'useLiveUpload config.accept must be "any" or a non-empty array of strings',
    );
  }

  if (new Set(value).size !== value.length) {
    throw new TypeError(
      "useLiveUpload config.accept must not contain duplicates",
    );
  }
  return Object.freeze([...value]);
}

function normalizeConfigErrors(
  value: unknown,
  allowedRefs: ReadonlySet<string>,
): readonly NormalizedLiveUploadError[] {
  if (!Array.isArray(value)) {
    throw new TypeError("useLiveUpload config.errors must be an array");
  }

  return Object.freeze(
    value.map((error, index) => {
      const source = `useLiveUpload config.errors[${index}]`;
      if (!isRecord(error)) {
        throw new TypeError(`${source} must be an object`);
      }
      assertExactKeys(error, CONFIG_ERROR_KEYS, source);
      const ref = assertNonEmptyString(error.ref, `${source}.ref`);
      if (!allowedRefs.has(ref)) {
        throw new TypeError(`${source}.ref does not identify this upload`);
      }
      return Object.freeze({
        error: cloneEncodedValue(error.error, `${source}.error`),
        ref,
      });
    }),
  );
}

export function normalizeLiveUploadConfig(
  config: unknown,
): NormalizedLiveUploadConfig {
  if (!isRecord(config)) {
    throw new TypeError(
      "useLiveUpload requires an encoded Phoenix upload config object",
    );
  }
  assertExactKeys(config, CONFIG_KEYS, "useLiveUpload config");

  if (!Array.isArray(config.entries)) {
    throw new TypeError("useLiveUpload config.entries must be an array");
  }
  const entries = Object.freeze(
    config.entries.map((entry, index) =>
      normalizeEntry(entry, `useLiveUpload config.entries[${index}]`),
    ),
  );
  const refs = entries.map(({ ref }) => ref);
  if (new Set(refs).size !== refs.length) {
    throw new TypeError("useLiveUpload config.entries must have unique refs");
  }

  const maxEntriesMode = config.max_entries_mode;
  if (maxEntriesMode !== "selected" && maxEntriesMode !== "total") {
    throw new TypeError(
      'useLiveUpload config.max_entries_mode must be "selected" or "total"',
    );
  }
  const ref = assertNonEmptyString(config.ref, "useLiveUpload config.ref");
  const allowedErrorRefs = new Set([ref, ...refs]);

  return Object.freeze({
    accept: normalizeAccept(config.accept),
    auto_upload: assertBoolean(
      config.auto_upload,
      "useLiveUpload config.auto_upload",
    ),
    entries,
    errors: normalizeConfigErrors(config.errors, allowedErrorRefs),
    max_entries: assertSafeInteger(
      config.max_entries,
      "useLiveUpload config.max_entries",
      1,
    ),
    max_entries_mode: maxEntriesMode,
    max_file_size: assertSafeInteger(
      config.max_file_size,
      "useLiveUpload config.max_file_size",
      1,
    ),
    name: assertNonEmptyString(config.name, "useLiveUpload config.name"),
    ref,
  });
}

function isElement(value: unknown): value is HTMLElement {
  if (
    typeof value !== "object" ||
    value === null ||
    !("ownerDocument" in value)
  ) {
    return false;
  }
  const ownerDocument = (value as { readonly ownerDocument?: Document })
    .ownerDocument;
  const ownerWindow = ownerDocument?.defaultView;
  return (
    ownerWindow !== null &&
    ownerWindow !== undefined &&
    value instanceof ownerWindow.HTMLElement
  );
}

function resolveTargetElement(
  target: LiveViewTarget | undefined,
  ownerDocument: Document,
  input: HTMLInputElement,
): HTMLElement | null {
  if (target === undefined) return null;
  if (typeof target === "number") {
    const selector = `[data-phx-component="${target}"]`;
    const component =
      input.closest<HTMLElement>(selector) ??
      input.form?.closest<HTMLElement>(selector);
    if (component === null || component === undefined) {
      throw new Error(
        `useLiveUpload input "${input.id}" is outside component target ${target}`,
      );
    }
    return component;
  }
  if (isElement(target)) {
    if (target.ownerDocument !== ownerDocument) {
      throw new Error("useLiveUpload target belongs to another document");
    }
    return target;
  }

  let elements: readonly Element[];
  try {
    elements = Array.from(ownerDocument.querySelectorAll(target));
  } catch (error) {
    throw new TypeError(
      `useLiveUpload target is not a valid selector: ${String(error)}`,
    );
  }
  if (elements.length === 0) {
    throw new Error(`useLiveUpload could not find target selector "${target}"`);
  }
  if (elements.length !== 1 || !isElement(elements[0])) {
    throw new Error(
      `useLiveUpload target selector "${target}" must match exactly one element`,
    );
  }
  const element = elements[0];
  return element;
}

function phoenixEvent(form: HTMLFormElement, kind: "change" | "submit") {
  const direct = form.getAttribute(`phx-${kind}`);
  const data = form.getAttribute(`data-phx-${kind}`);
  if (direct !== null && data !== null && direct !== data) {
    throw new Error(
      `useLiveUpload form has conflicting phx-${kind} attributes`,
    );
  }
  return direct ?? data;
}

function assertExpectedFormEvent(
  form: HTMLFormElement,
  kind: "change" | "submit",
  expected: string | undefined,
): void {
  if (expected === undefined) return;
  const actual = phoenixEvent(form, kind);
  if (actual !== expected) {
    throw new Error(
      `useLiveUpload expected form phx-${kind}="${expected}", got ${
        actual === null ? "no binding" : `"${actual}"`
      }`,
    );
  }
}

function uploadInputsInScope(
  ownerDocument: Document,
  scope: HTMLElement,
): readonly HTMLInputElement[] {
  return Array.from(
    ownerDocument.querySelectorAll<HTMLInputElement>(
      'input[type="file"][data-phx-upload-ref]',
    ),
  ).filter((candidate) => {
    if (scope instanceof ownerDocument.defaultView!.HTMLFormElement) {
      return candidate.form === scope;
    }
    return (
      (scope !== candidate && scope.contains(candidate)) ||
      (scope.id !== "" && candidate.getAttribute("form") === scope.id)
    );
  });
}

export function resolveLiveUploadInput(
  config: NormalizedLiveUploadConfig,
  options: ResolveLiveUploadInputOptions,
): ResolvedLiveUploadInput {
  const ownerDocument =
    options.bridgeElement?.ownerDocument ??
    (typeof document === "undefined" ? null : document);
  if (ownerDocument === null || ownerDocument.defaultView === null) {
    throw new Error("useLiveUpload requires a browser document");
  }

  const candidate = ownerDocument.getElementById(config.ref);
  if (
    !(candidate instanceof ownerDocument.defaultView.HTMLInputElement) ||
    candidate.type !== "file"
  ) {
    throw new Error(
      `useLiveUpload could not find a live_file_input with id "${config.ref}"`,
    );
  }
  if (candidate.getAttribute("data-phx-hook") !== "Phoenix.LiveFileUpload") {
    throw new Error(
      `useLiveUpload input "${config.ref}" is not rendered by <.live_file_input>`,
    );
  }
  if (candidate.getAttribute("data-phx-upload-ref") !== config.ref) {
    throw new Error(
      `useLiveUpload input "${config.ref}" has a mismatched upload ref`,
    );
  }
  if (candidate.name !== config.name) {
    throw new Error(
      `useLiveUpload input "${config.ref}" must have name "${config.name}"`,
    );
  }
  if (candidate.multiple !== config.max_entries > 1) {
    throw new Error(
      `useLiveUpload input "${config.ref}" has a multiple attribute inconsistent with max_entries`,
    );
  }
  const expectedAccept = config.accept === "any" ? "" : config.accept.join(",");
  if (candidate.accept !== expectedAccept) {
    throw new Error(
      `useLiveUpload input "${config.ref}" has an accept attribute inconsistent with config.accept`,
    );
  }
  if (candidate.hasAttribute("data-phx-auto-upload") !== config.auto_upload) {
    throw new Error(
      `useLiveUpload input "${config.ref}" has an auto-upload attribute inconsistent with config.auto_upload`,
    );
  }

  const form = candidate.form;
  if (form === null) {
    throw new Error(
      `useLiveUpload input "${config.ref}" must be associated with a form`,
    );
  }
  if (options.formId !== undefined && form.id !== options.formId) {
    throw new Error(
      `useLiveUpload input "${config.ref}" must be associated with form "${options.formId}"`,
    );
  }

  assertExpectedFormEvent(form, "change", options.changeEvent);
  assertExpectedFormEvent(form, "submit", options.submitEvent);
  if (config.auto_upload && phoenixEvent(form, "change") === null) {
    throw new Error(
      `useLiveUpload auto-upload input "${config.ref}" requires a form phx-change binding`,
    );
  }

  const explicitTarget = resolveTargetElement(
    options.target,
    ownerDocument,
    candidate,
  );
  const liveViewOwner =
    options.bridgeElement?.closest<HTMLElement>("[data-phx-session]");
  const scope = explicitTarget ?? liveViewOwner ?? form;
  const matchingInputs = uploadInputsInScope(ownerDocument, scope).filter(
    (input) => input.name === config.name,
  );
  if (!matchingInputs.includes(candidate)) {
    throw new Error(
      `useLiveUpload input "${config.ref}" is outside the upload target`,
    );
  }
  if (matchingInputs.length !== 1) {
    throw new Error(
      `useLiveUpload found duplicate live file inputs named "${config.name}"`,
    );
  }

  return Object.freeze({ form, input: candidate });
}

export function normalizeUploadFiles(
  files: UploadFiles | null | undefined,
): readonly File[] {
  if (files === null || files === undefined) return Object.freeze([]);
  if (typeof File === "undefined") {
    throw new Error("useLiveUpload addFiles requires a browser File API");
  }

  const normalized = Array.from(files);
  if (!normalized.every((file) => file instanceof File)) {
    throw new TypeError("useLiveUpload addFiles accepts only File objects");
  }
  return Object.freeze(normalized);
}

File: lib/live_view_react/uploads.ex
defmodule LiveViewReact.Uploads do
  @moduledoc false

  alias LiveViewReact.Encoder
  alias Phoenix.LiveView.{UploadConfig, UploadEntry}
  @max_progress 100

  @spec encode_config(UploadConfig.t(), keyword()) :: map()
  def encode_config(%UploadConfig{} = config, opts) do
    base_entries =
      Enum.map(config.entries, fn entry ->
        encode_entry(entry, opts)
      end)

    entry_refs = MapSet.new(base_entries, & &1.ref)
    errors = encode_errors(config.errors, config.ref, entry_refs, opts)

    entries =
      Enum.map(base_entries, fn entry ->
        Map.put(entry, :errors, errors_for_entry(errors, entry.ref))
      end)

    Encoder.encode(
      %{
        accept: encode_accept(config.accept),
        auto_upload: config.auto_upload?,
        entries: entries,
        errors: errors,
        max_entries: validate_positive_integer!(config.max_entries, "max_entries"),
        max_entries_mode: encode_max_entries_mode(config.max_entries_mode),
        max_file_size: validate_positive_integer!(config.max_file_size, "max_file_size"),
        name: encode_name(config.name),
        ref: validate_non_empty_binary!(config.ref, "ref")
      },
      opts
    )
  end

  @spec encode_entry(UploadEntry.t(), keyword()) :: map()
  def encode_entry(%UploadEntry{} = entry, opts) do
    progress = validate_progress!(entry.progress)

    Encoder.encode(
      %{
        cancelled: entry.cancelled?,
        client_last_modified:
          validate_non_negative_integer!(
            entry.client_last_modified || 0,
            "client_last_modified"
          ),
        client_name: validate_binary!(entry.client_name, "client_name"),
        client_relative_path: entry.client_relative_path || "",
        client_size: validate_non_negative_integer!(entry.client_size, "client_size"),
        client_type: validate_binary!(entry.client_type || "", "client_type"),
        done: entry.done?,
        errors: [],
        preflighted: entry.preflighted?,
        progress: progress,
        ref: validate_non_empty_binary!(entry.ref, "entry ref"),
        valid: entry.valid?
      },
      opts
    )
  end

  defp encode_accept(:any), do: "any"

  defp encode_accept(accept) when is_binary(accept) do
    accept
    |> String.split(",", trim: true)
    |> Enum.map(&String.trim/1)
    |> validate_accept_list!()
  end

  defp encode_accept(accept) when is_list(accept),
    do: validate_accept_list!(accept)

  defp encode_accept(accept) do
    raise ArgumentError, "unsupported upload accept contract: #{inspect(accept)}"
  end

  defp encode_max_entries_mode(:selected), do: "selected"
  defp encode_max_entries_mode(:total), do: "total"
  defp encode_max_entries_mode("selected"), do: "selected"
  defp encode_max_entries_mode("total"), do: "total"

  defp encode_max_entries_mode(mode) do
    raise ArgumentError,
          "upload max_entries_mode must be :selected or :total, got: #{inspect(mode)}"
  end

  defp encode_name(name) when is_atom(name),
    do: validate_non_empty_binary!(Atom.to_string(name), "name")

  defp encode_name(name) when is_binary(name),
    do: validate_non_empty_binary!(name, "name")

  defp encode_name(name) do
    raise ArgumentError, "upload name must be an atom or binary, got: #{inspect(name)}"
  end

  defp encode_errors(errors, config_ref, entry_refs, opts) do
    Enum.map(errors, fn {ref, reason} ->
      validate_error_ref!(ref, config_ref, entry_refs)

      %{
        error: Encoder.encode(reason, opts),
        ref: ref
      }
    end)
  end

  defp errors_for_entry(errors, entry_ref) do
    for %{error: error, ref: ^entry_ref} <- errors, do: error
  end

  defp validate_accept_list!(accept) do
    normalized =
      Enum.map(accept, fn entry ->
        entry
        |> validate_binary!("accept entry")
        |> String.trim()
      end)

    cond do
      normalized == [] ->
        raise ArgumentError, "upload accept list must not be empty"

      Enum.any?(normalized, &(&1 == "")) ->
        raise ArgumentError, "upload accept entries must be non-empty strings"

      MapSet.size(MapSet.new(normalized)) != length(normalized) ->
        raise ArgumentError, "upload accept entries must be unique"

      true ->
        normalized
    end
  end

  defp validate_error_ref!(ref, config_ref, entry_refs) do
    ref = validate_non_empty_binary!(ref, "upload error ref")

    if ref == config_ref or MapSet.member?(entry_refs, ref) do
      ref
    else
      raise ArgumentError,
            "upload error ref must match the config ref or an encoded entry ref, got: #{inspect(ref)}"
    end
  end

  defp validate_progress!(progress)
       when is_integer(progress) and progress >= 0 and progress <= @max_progress,
       do: progress

  defp validate_progress!(progress) do
    raise ArgumentError,
          "upload progress must be an integer from 0 through #{@max_progress}, got: #{inspect(progress)}"
  end

  defp validate_positive_integer!(value, _field)
       when is_integer(value) and value > 0,
       do: value

  defp validate_positive_integer!(value, field) do
    raise ArgumentError, "upload #{field} must be a positive integer, got: #{inspect(value)}"
  end

  defp validate_non_negative_integer!(value, _field)
       when is_integer(value) and value >= 0,
       do: value

  defp validate_non_negative_integer!(value, field) do
    raise ArgumentError,
          "upload #{field} must be a non-negative integer, got: #{inspect(value)}"
  end

  defp validate_non_empty_binary!(value, _field)
       when is_binary(value) and value != "",
       do: value

  defp validate_non_empty_binary!(value, field) do
    raise ArgumentError, "upload #{field} must be a non-empty string, got: #{inspect(value)}"
  end

  defp validate_binary!(value, _field) when is_binary(value), do: value

  defp validate_binary!(value, field) do
    raise ArgumentError, "upload #{field} must be a string, got: #{inspect(value)}"
  end
end
